Legacy Audio Bootstrapping

capnkenny opened the pull request:
This should bootstrap the current audio impl. to allow it to work under ECS.

Note: in no way will this be pluginified in its current shape unless we decide not to revamp audio later on. (It could be, but I wanted to settle on a better solution instead of solely relying on OpenAL)

Resolves #346

RubyNova commented on include/NovelRT.Interop/Audio/NrtAudio.h:
```diff
@@ -0,0 +1,9 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#ifndef NOVELRT_INTEROP_AUDIO_H
+#define NOVELRT_INTEROP_AUDIO_H
+
+#include "NrtAudioService.h"
+
+#endif // NOVELRT_INTEROP_AUDIO_H
```
Empty header? Forget to add headers here?

capnkenny replied:
This is not strictly empty, but it has been moved to match how `NrtEcs.h` is formatted and utilised. Please let me know if you'd expect something different here? Even though there's only one header being included, I thought we may rather keep consistency here instead of changing it depending on the namespace.

RubyNova replied:
Oh, I misread it in that case. I don't remember seing any includes previously.

RubyNova commented on include/NovelRT.Interop/Audio/NrtAudioService.h:
```diff
@@ -37,6 +37,8 @@ extern "C"
     NrtResult Nrt_AudioService_playSound(NrtAudioServiceHandle service, unsigned int handle, int loops);
     NrtResult Nrt_AudioService_stopSound(NrtAudioServiceHandle service, unsigned int handle);
     NrtResult Nrt_AudioService_tearDown(NrtAudioServiceHandle service);
+    NrtBool Nrt_AudioService_isMusicLoaded(NrtAudioServiceHandle service, NrtAudioServiceIteratorHandle handle);
+    NrtBool Nrt_AudioService_isSoundLoaded(NrtAudioServiceHandle service, unsigned int handle);
```
Is the casing due to the fact this is a legacy API?

capnkenny replied:
Yeppers

capnkenny replied:
Switched casing to be more consistent with modern APIs.

RubyNova commented on include/NovelRT.Interop/Ecs/Audio/NrtAudioEmitterComponent.h:
```diff
@@ -0,0 +1,19 @@
+{
+#endif
+
+    void Nrt_AudioEmitterComponent_Update(void* lhs, const void* rhs, size_t size, void* context);
```
Is there a reason the component pointers are `void*` if the component type is strongly known?

capnkenny replied:
Thought I needed void*s due to signature, but I was mistaken. Resolved.

RubyNova commented on include/NovelRT.Interop/Ecs/Audio/NrtEcsAudioTypedefs.h:
```diff
@@ -0,0 +1,45 @@
+    typedef enum
+    {
+        NRT_EMITTER_STATE_DONE = 0,
+        NRT_EMITTER_STATE_TOPLAY = 1,
```
Did you mean TO_PLAY?

capnkenny replied:
Corrected

RubyNova commented on include/NovelRT.Interop/Ecs/Audio/NrtEcsAudioTypedefs.h:
```diff
@@ -0,0 +1,45 @@
+        NRT_EMITTER_STATE_PLAYING = 2,
+        NRT_EMITTER_STATE_TOSTOP = 3,
+        NRT_EMITTER_STATE_STOPPED = 4,
+        NRT_EMITTER_STATE_TOPAUSE = 5,
```
Same here?

capnkenny replied:
Same here

RubyNova commented on include/NovelRT.Interop/Ecs/Audio/NrtEcsAudioTypedefs.h:
```diff
@@ -0,0 +1,45 @@
+        NRT_EMITTER_STATE_STOPPED = 4,
+        NRT_EMITTER_STATE_TOPAUSE = 5,
+        NRT_EMITTER_STATE_PAUSED = 6,
+        NRT_EMITTER_STATE_TORESUME = 7,
```
What about here?

capnkenny replied:
Samesies

RubyNova commented on include/NovelRT.h:
```diff
@@ -187,6 +173,9 @@
   #include "NovelRT/Animation/SpriteAnimator.h"
 
+// Audio types
+ #include "NovelRT/Audio/Audio.h"
```
Spacing/indentation looks a little wonky here?

capnkenny replied:
Haven't applied clang-format yet, and I may have spaced a little weird 😅 still getting used to CLion

capnkenny replied:
Manually resolved for now

RubyNova commented on include/NovelRT/Audio/Audio.h:
```diff
@@ -0,0 +1,33 @@
+#include <sndfile.h>
+//#include "../Maths/GeoVector2F.h"
+
+namespace NovelRT::Audio
+{
+    class AudioService;
+}
```
This looks to be living outside of the C++ check? I'm not sure I 100% understand what's happening here.

capnkenny replied:
This appears to be what we are doing with `Ecs.h`, `Graphics.h`, `Windowing.h`, etc.. so I was just following the semantic there.

RubyNova replied:
`Ecs.h` has no C API calls. The header is excluded via `NovelRT.h` if the C API is used, I'm fairly sure? I might be misunderstanding something here though. This sounds like a possible bug vector otherwise.

capnkenny replied:
Odd, I thought we had this for a reason (it was in NovelRT.h at some point im pretty sure). I've removed it in a new commit, compiles fine ¯\_(ツ)_/¯

RubyNova commented on include/NovelRT/Audio/Audio.h:
```diff
@@ -0,0 +1,33 @@
+#include <AL/al.h>
+#include <AL/alc.h>
+#include <sndfile.h>
+//#include "../Maths/GeoVector2F.h"
```
Dead code.

capnkenny replied:
Corrected in new commit.

RubyNova commented on include/NovelRT/Ecs/Audio/AudioEmitterStateComponent.h:
```diff
@@ -0,0 +1,25 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#ifndef NOVELRT_ECS_AUDIO_AUDIOEMITTERSTATECOMPONENT_H
+#define NOVELRT_ECS_AUDIO_AUDIOEMITTERSTATECOMPONENT_H
+
```
Are we ignoring the namespace error guard for this namespace since its legacy?

capnkenny replied:
No, this should have a guard akin to the ECS ones even if it is legacy. i'll fix it

RubyNova commented on resources/CMakeLists.txt:
```diff
@@ -16,6 +16,10 @@ set(RESOURCES_FILES
   Shaders/vert.spv
+
+  Sounds/waltz.ogg
+  Sounds/lazer.ogg
```
UK spelling please (or at least the meme spelling :D )

capnkenny replied:
Yep, I messed up memes - will fix, very disappoint

capnkenny replied:
Done

RubyNova commented on samples/Ecs/AudioEcsSample/main.cpp:
```diff
@@ -0,0 +1,230 @@
+
+using namespace NovelRT::Ecs::Audio;
+
+int main(int /*argc*/, char* /*argv*/[])
```
You could probably just delete the parameters here. I don't think main requires you have them, esp. if you're not using them?

capnkenny replied:
Done

RubyNova commented on samples/Ecs/AudioEcsSample/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+  COMMAND ${CMAKE_COMMAND} -E copy_directory
+    $<TARGET_FILE_DIR:Engine>
+    $<TARGET_FILE_DIR:AudioEcsSample>
+)
```
missing EOF.

RubyNova commented on samples/Ecs/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+add_subdirectory(AudioEcsSample)
+add_subdirectory(AudioEcsCAPI)
```
Missing EOF.

RubyNova commented on samples/Ecs/AudioEcsSample/main.cpp:
```diff
@@ -0,0 +1,230 @@
+    runner.Update += [&](NovelRT::Timing::Timestamp delta) { scheduler.ExecuteIteration(delta); };
+
+    runner.SceneConstructionRequested += [&] {
+        //playMusicButton->executeObjectBehaviour();
+        fadeMusicInButton->executeObjectBehaviour();
+        fadeMusicOutButton->executeObjectBehaviour();
+        playLazerButton->executeObjectBehaviour();
+        playGoatButton->executeObjectBehaviour();
+        //playMusicText->executeObjectBehaviour();
+        fadeMusicInText->executeObjectBehaviour();
+        fadeMusicOutText->executeObjectBehaviour();
+        playLazerText->executeObjectBehaviour();
+        playGoatText->executeObjectBehaviour();
+        //playMusicInteraction->executeObjectBehaviour();
```
Dead code in the comments here maybe?

capnkenny replied:
Removed

RubyNova commented on samples/Ecs/AudioEcsSample/main.cpp:
```diff
@@ -0,0 +1,230 @@
+        playGoatInteraction->executeObjectBehaviour();
+    };
+
+    // GASSSSSSSSS GASSSSSSSSSS GASSSSSSSSSSSSSS
```
🚗 🚗 🚗 🚗 🚗 🚗 🚗 🚗

capnkenny replied:
Removed memery - much meme, very wow

RubyNova commented on samples/Experimental/VulkanRender/CMakeLists.txt:
```diff
@@ -3,29 +3,12 @@ set(SOURCES
 
 add_executable(VulkanRender ${SOURCES})
-add_dependencies(VulkanRender Dotnet Resources)
+add_dependencies(VulkanRender Resources)
```
Uhhhhhhhhh.........why was this modified in this PR? Am I missing context?

capnkenny replied:
It was a missed change in another PR that was blocking me lol. I could have separated it but I didn't

RubyNova replied:
How is VulkanRender blocking you?

capnkenny replied:
Should be reverted as well

RubyNova commented on samples/Experimental/VulkanRender/CMakeLists.txt:
```diff
@@ -3,29 +3,12 @@ set(SOURCES
         Engine
         )
 
-add_custom_command(
-        TARGET VulkanRender POST_BUILD
-        COMMAND ${CMAKE_COMMAND} -E copy_directory
-        $<TARGET_FILE_DIR:Dotnet>
-        $<TARGET_FILE_DIR:VulkanRender>/dotnet
-)
-
-#this is pure hacky hotfix goodness. We need to figure out a better way to do this in the future.
-if(WIN32)
-    add_custom_command(
-            TARGET VulkanRender POST_BUILD
-            COMMAND ${CMAKE_COMMAND} -E copy
-            $<TARGET_FILE_DIR:VulkanRender>/dotnet/nethost.dll
-            $<TARGET_FILE_DIR:VulkanRender>
-    )
-endif()
-
```
Much confuse why VulkanRender is being touched here.

capnkenny replied:
Shouldn't be including Ink when not required - this breaks configuring if ink is defined to _not_ be included. Again, probably should be removed and moved to another Pr, but idk how much pain that will be

RubyNova replied:
I'd much rather this was in another PR. It's adding noise. Are these changes required for this PR to work?

capnkenny replied:
Reverted

RubyNova commented on src/NovelRT.Interop/CMakeLists.txt:
```diff
@@ -19,12 +19,16 @@ set(SOURCES
   Ecs/NrtSystemScheduler.cpp
   Ecs/NrtUnsafeComponentView.cpp
 
+        Ecs/Audio/NrtAudioSystem.cpp
+        Ecs/Audio/NrtAudioEmitterComponent.cpp
+        Ecs/Audio/NrtAudioEmitterStateComponent.cpp
+
```
Indentation issues here?

capnkenny replied:
🌊🦁

RubyNova commented on src/NovelRT.Interop/Ecs/Audio/NrtAudioEmitterComponent.cpp:
```diff
@@ -0,0 +1,21 @@
+
+#ifdef __cplusplus
+}
+#endif
```
Missing EOF.

RubyNova commented on src/NovelRT.Interop/Ecs/Audio/NrtAudioEmitterStateComponent.cpp:
```diff
@@ -0,0 +1,21 @@
+
+#ifdef __cplusplus
+}
+#endif
```
Missing EOF.

RubyNova commented on src/NovelRT.Interop/Ecs/NrtComponentCache.cpp:
```diff
@@ -63,8 +63,11 @@ extern "C"
         }
 
+        auto test = reinterpret_cast<ComponentCache*>(componentCache);
+        auto lol = test->GetComponentBufferById(id).get();
+
         *outputResult = reinterpret_cast<NrtComponentBufferMemoryContainerHandle>(
-            reinterpret_cast<ComponentCache*>(componentCache)->GetComponentBufferById(id).get());
+            lol);
```
These names are really confusing. Is this debug code? Unit test code?

capnkenny replied:
Oh no, ill revert this! Just was test code

capnkenny replied:
Fixed in newer commit

RubyNova commented on src/NovelRT.Interop/NrtNovelRunner.cpp:
```diff
@@ -66,34 +65,6 @@ extern "C"
         return NRT_SUCCESS;
     }
 
-    NrtResult Nrt_NovelRunner_getAudioService(NrtNovelRunnerHandle runner, NrtAudioServiceHandle* outputService)
-    {
-        if (runner == nullptr)
-        {
-            Nrt_setErrMsgIsNullptrInternal();
-            return NRT_FAILURE_NULL_INSTANCE_PROVIDED;
-        }
-
-        if (outputService == nullptr)
-        {
-            Nrt_setErrMsgIsNullptrInternal();
-            return NRT_FAILURE_NULL_ARGUMENT_PROVIDED;
-        }
-
-        NovelRunner* cRunner = reinterpret_cast<NovelRunner*>(runner);
-        _audioCollection.push_back(cRunner->getAudioService());
-
-        auto ptr = _audioCollection.back().get();
-        if (ptr == nullptr)
-        {
-            return NRT_FAILURE_NULL_INSTANCE_PROVIDED;
-        }
-
-        *outputService = reinterpret_cast<NrtAudioServiceHandle>(ptr);
-
-        return NRT_SUCCESS;
-    }
-
```
Is this part of our efforts to move away from the legacy NovelRunner?

capnkenny replied:
Yep, and is required since it won't map to anythjng in C++ land

RubyNova commented on src/NovelRT/Audio/AudioService.cpp:
```diff
@@ -2,6 +2,7 @@
 // for more information.
 
 #include <NovelRT.h>
+#include <NovelRT/Audio/AudioService.h>
```
I'd probably remove the NovelRT.h include here unless you absolutely need it. Seems redundant to have both.

capnkenny replied:
It's legacy code, but I can adjust it since we're not dependent on novelrunner anymore

capnkenny replied:
Done

RubyNova commented on include/NovelRT/Audio/AudioService.h:
```diff
@@ -57,7 +53,11 @@ namespace NovelRT::Audio
         void playSound(ALuint handle, int32_t loops);
         void stopSound(ALuint handle);
         void tearDown();
+        bool isLoaded(std::vector<ALuint>::iterator handle);
+        bool isLoaded(ALuint handle);
+        bool isMusicPlaying();
+        bool isSoundPlaying(ALuint handle);
```
these should all be `[[nodiscard]]`.

capnkenny replied:
done in new commit

RubyNova commented on include/NovelRT/Ecs/Audio/Ecs.Audio.h:
```diff
@@ -0,0 +1,29 @@
+#include "AudioEmitterStateComponent.h"
+#include "AudioSystem.h"
+
+#endif // NOVELRT_ECS_AUDIO_H
```
Missing EOF.

capnkenny replied:
I've been afflicted by the sea lion that gets you too, methinks 🤔

capnkenny commented on CMakeLists.txt:
```diff
@@ -15,6 +15,9 @@ if(MSVC)
   string(REGEX REPLACE " /W[0-4]" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
   string(REGEX REPLACE " /M[TD]d?" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
   string(REGEX REPLACE " /M[TD]d?" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+  #May need to set the DEBUG version of these as well if someone uses the VS Generator
+  set(CMAKE_SHARED_LINKER_FLAGS "/INCREMENTAL:YES ${CMAKE_SHARED_LINKER_FLAGS}")
+  set(CMAKE_EXE_LINKER_FLAGS  "/INCREMENTAL:YES ${CMAKE_EXE_LINKER_FLAGS}")
```
Testing incremental linking under MSVC - this doesn't really fix the copying issue we face with CLion and VS 2019, but it does help with additional builds (when the copying works)

capnkenny replied:
Removed in newest commit

capnkenny commented on include/NovelRT.Interop/Ecs/Audio/NrtEcsAudioTypedefs.h:
```diff
@@ -0,0 +1,45 @@
+        NRT_EMITTER_STATE_DONE = 0,
+        NRT_EMITTER_STATE_TOPLAY = 1,
+        NRT_EMITTER_STATE_PLAYING = 2,
+        NRT_EMITTER_STATE_TOSTOP = 3,
```
Corrected this in newer commit

capnkenny commented on include/NovelRT.Interop/Ecs/Audio/NrtAudioEmitterStateComponent.h:
```diff
@@ -0,0 +1,19 @@
+{
+#endif
+
+    void Nrt_AudioEmitterStateComponent_Update(void* lhs, const void* rhs, size_t size, void* context);
```
Thought I needed void*s due to signature, but I was mistaken. Resolved.

capnkenny commented:
One more review for the homies? (hopefully lol) @RubyNova

RubyNova commented on include/NovelRT.Interop/Audio/NrtAudioService.h:
```diff
@@ -11,32 +11,38 @@ extern "C"
+    NrtResult Nrt_AudioService_LoadMusic(NrtAudioServiceHandle service,
                                          char* input,
                                          NrtAudioServiceIteratorHandle* output);
 
-    NrtResult Nrt_AudioService_setSoundVolume(NrtAudioServiceHandle service, unsigned int source, float val);
-    NrtResult Nrt_AudioService_setSoundPosition(NrtAudioServiceHandle service,
+    NrtResult Nrt_AudioService_SetSoundVolume(NrtAudioServiceHandle service, unsigned int source, float val);
+    NrtResult Nrt_AudioService_SetSoundPosition(NrtAudioServiceHandle service,
                                                 unsigned int source,
                                                 float posX,
                                                 float posY);
-    NrtResult Nrt_AudioService_resumeMusic(NrtAudioServiceHandle service);
-    NrtResult Nrt_AudioService_playMusic(NrtAudioServiceHandle service,
+    NrtResult Nrt_AudioService_ResumeMusic(NrtAudioServiceHandle service);
+    NrtResult Nrt_AudioService_PlayMusic(NrtAudioServiceHandle service,
                                          NrtAudioServiceIteratorHandle handle,
                                          int32_t loops);
```
mega nit: Inconsistent spacing between the signatures here. I'm guessing clang-format did this? Looks weird either way. Don't spend too long on it if clang-format refuses to cooperate though.

capnkenny replied:
Yup, this was 100% clang-format as we've set it up to do this, however are you referring to the parameter spacing here?

RubyNova replied:
No, I'm referring to the fact that multi-line signatures have no spacing between each other, like:
```cpp
MyThing(A a,
        B b);
MyOtherThing(A a,
             B b);
```

As opposed to:
```cpp
MyThing(A a,
        B b);
        
MyOtherThing(A a,
             B b);
```

capnkenny replied:
Not sure - could've been a side effect that wasn't manually resolved.

Cleaned it up now

RubyNova commented on include/NovelRT.Interop/Ecs/Audio/NrtAudioSystem.h:
```diff
@@ -0,0 +1,32 @@
+{
+#endif
+
+    NrtAudioSystemHandle Nrt_AudioSystem_Create();
```
I might be just being blind here, but is there an appropriate destroy/delete function for this?

capnkenny replied:
Noooooooo :^) added it in the next commits

RubyNova commented on include/NovelRT/Audio/Audio.h:
```diff
@@ -0,0 +1,33 @@
+#include <AL/alc.h>
+#include <cstdint>
+#include <sndfile.h>
+#include <vector>
```
This won't compile if in C API land. Not sure how this is currently passing. I think we need to review the structure of your C API flag here, because I'm becoming more and more confused the more I look at this file. 😅

capnkenny replied:
This shouldn't need to be compiled in C API land - it's a C++ header 😄 

The call that you see afterwards is only if c++ is defined and not NOVELRT_C_API - this is only used for C# generation AFAIK. C API will strictly use the Interop headers (where this doesn't exist).
I'll still play around with it, but removing the section that used to exist for C# generation might cause it to break due to the underlying types being typedef'ed

capnkenny replied:
Yeah, its odd but NovelRT.h seems to handle most of the C_API flag stuff, so I've just removed it. That way, if we're in C API, this isn't being compiled but the Interop Header is.

RubyNova commented on resources/CMakeLists.txt:
```diff
@@ -16,6 +16,10 @@ set(RESOURCES_FILES
   Shaders/vert.spv
+
+  Sounds/waltz.ogg
+        Sounds/laz0r.ogg
```
nit: indentation issue.

capnkenny replied:
BAH CLion -_-. Fixed

RubyNova commented on samples/Ecs/AudioEcsCAPI/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+                $<TARGET_FILE_DIR:AudioEcsCAPI>
+        )
+    endif()
+endif()
```
Missing EOF.

capnkenny replied:
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

RubyNova commented on samples/Ecs/AudioEcsCAPI/main.c:
```diff
@@ -0,0 +1,287 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#define NOVELRT_C_API 1
+#include <NovelRT.Interop/Ecs/Audio/NrtEcsAudio.h>
+#include <NovelRT.h>
```
Shouldn't you only need NovelRT.h here since you're defining the C API flag for it? What am I misunderstanding?

capnkenny replied:
Probably  - looks like missing headers for NovelRT.h

RubyNova commented on src/NovelRT.DotNet/Interop/Audio/NovelRT.cs:
```diff
@@ -9,54 +9,54 @@ public static unsafe partial class NovelRT
-        public static extern NrtResult Nrt_AudioService_playMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("NrtAudioServiceIteratorHandle")] IntPtr handle, [NativeTypeName("int32_t")] int loops);
+        public static extern NrtResult Nrt_AudioService_PlayMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("NrtAudioServiceIteratorHandle")] IntPtr handle, [NativeTypeName("int32_t")] int loops);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_pauseMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
+        public static extern NrtResult Nrt_AudioService_PauseMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_stopMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
+        public static extern NrtResult Nrt_AudioService_StopMusic([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_setMusicVolume([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, float value);
+        public static extern NrtResult Nrt_AudioService_SetMusicVolume([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, float value);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_checkSources([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
+        public static extern NrtResult Nrt_AudioService_CheckSources([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_loadSound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("char *")] sbyte* input, [NativeTypeName("unsigned int *")] uint* output);
+        public static extern NrtResult Nrt_AudioService_LoadSound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("char *")] sbyte* input, [NativeTypeName("unsigned int *")] uint* output);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_unload([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle);
+        public static extern NrtResult Nrt_AudioService_Unload([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_playSound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle, int loops);
+        public static extern NrtResult Nrt_AudioService_PlaySound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle, int loops);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_stopSound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle);
+        public static extern NrtResult Nrt_AudioService_StopSound([NativeTypeName("NrtAudioServiceHandle")] IntPtr service, [NativeTypeName("unsigned int")] uint handle);
 
         [DllImport("NovelRT.Interop", ExactSpelling = true)]
-        public static extern NrtResult Nrt_AudioService_tearDown([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
+        public static extern NrtResult Nrt_AudioService_TearDown([NativeTypeName("NrtAudioServiceHandle")] IntPtr service);
     }
 }
```
Oh wow, nice! I forgot we needed this! Good catch :D

capnkenny replied:
Thank CLion lol

RubyNova commented on src/NovelRT.Interop/Audio/NrtAudioService.cpp:
```diff
@@ -333,7 +333,105 @@ extern "C"
+        return serv->IsSoundPlaying(handle) ? NRT_TRUE : NRT_FALSE;
+    }
+
+    NrtResult Nrt_AudioService_GetMusicVolume(NrtAudioServiceHandle service, float* output)
+    {
+        if (service == nullptr)
+        {
+            Nrt_setErrMsgIsNullptrInternal();
+            return NRT_FAILURE_NULL_INSTANCE_PROVIDED;
+        }
+
+        auto serv = reinterpret_cast<Audio::AudioService*>(service);
+
+        try
+        {
+            *output = serv->GetMusicVolume();
+        }
+        catch (const Exceptions::NotInitialisedException)
+        {
+            Nrt_setErrMsgIsNotInitialisedInternal();
+            return NRT_FAILURE_NOT_INITIALISED;
+        }
+
+        return NRT_SUCCESS;
+    }
```
Out of curiosity, is this getter this way because of the exception? I know most getters are straight up "grab the value or segfault"

capnkenny replied:
I can adjust this to segfault - I just realized that the audio service most likely wouldn't throw a Not Initialised exception here anyways

RubyNova replied:
If you have a look at some of our "high perf" getters, they just straight up return the value or segfault. However, if the getter can throw, then this is 100% correct. :)

RubyNova commented on include/NovelRT/Audio/AudioService.h:
```diff
@@ -26,38 +26,45 @@ namespace NovelRT::Audio
+        [[nodiscard]] bool IsLoaded(ALuint handle);
+        [[nodiscard]] bool IsMusicPlaying();
+        [[nodiscard]] bool IsSoundPlaying(ALuint handle);
+        float GetMusicVolume();
+        float GetSoundVolume(ALuint handle);
```
these should be [[nodiscard]] ideally. ☺️

capnkenny replied:
true, true

RubyNova reviewed (changes requested):
Minor nits and a concern over the C API flag still remain, but once we've resolved those I'll approve. Good stuff!

capnkenny commented:
@RubyNova should have fixed the latest round of everything requested. Also added the dtor for `AudioSystem` and added the equivalent Destroy function in Interop.

The only thing I cannot test is C# generation as I dunno how :^) Otherwise, it looks like samples are building and working

RubyNova commented:
You test it the same way you test C? I'm not sure I understand the confusion :P

RubyNova reviewed (approved):
There's one or two places we'd benefit from nodiscard attributes but honestly this is a temporary bootstrap and so I don't care enough. LGTM.
